metricbeat/module/prometheus/collector/data.go

```diff
@@ -1,83 +1,94 @@
 package collector
 
 import (
-	"strconv"
-	"strings"
-
 	"github.com/elastic/beats/libbeat/common"
+	dto "github.com/prometheus/client_model/go"
+	"math"
+	"strconv"
 )
 
 type PromEvent struct {
 	key       string
-	value     interface{}
+	value     common.MapStr
 	labels    common.MapStr
 	labelHash string
 }
 
-// NewPromEvent creates a prometheus event based on the given string
-func NewPromEvent(line string) PromEvent {
-	// Separate key and value
-	splitPos := strings.LastIndex(line, " ")
-	split := []string{line[:splitPos], line[splitPos+1:]}
-
-	promEvent := PromEvent{
-		key:       split[0],
-		labelHash: "_", // _ represents empty labels
-	}
-
-	// skip entries without a value
-	if split[1] == "NaN" {
-		promEvent.value = nil
-	} else {
-		promEvent.value = convertValue(split[1])
-	}
-
-	// Split key
-	startLabels := strings.Index(line, "{")
-	endLabels := strings.Index(line, "}")
-
-	// Handle labels
-	if startLabels != -1 {
-		// Overwrite key, as key contained labels until now too
-		promEvent.key = line[0:startLabels]
-		promEvent.labelHash = line[startLabels+1 : endLabels]
-		// Extract labels
-		promEvent.labels = extractLabels(promEvent.labelHash)
-	}
-
-	return promEvent
-}
-
-// extractLabels splits up a label string of format handler="alerts",quantile="0.5"
-// into a key / value list
-func extractLabels(labelsString string) common.MapStr {
-
-	keyValuePairs := common.MapStr{}
-
-	// Extract labels
-	labels := strings.Split(labelsString, "\",")
-	for _, label := range labels {
-		keyValue := strings.Split(label, "=")
-		// Remove " from value
-		keyValue[1] = strings.Trim(keyValue[1], "\"")
+func GetPromEventsFromMetricFamily(mf *dto.MetricFamily) []PromEvent {
+	var events []PromEvent
+
+	name := *mf.Name
+	metrics := mf.Metric
+	for _, metric := range metrics {
+		event := PromEvent{
+			key:       name,
+			labelHash: "#",
+		}
+		value := common.MapStr{}
+		labels := metric.Label
+
+		if len(labels) != 0 {
+			tagsMap := common.MapStr{}
+			for _, label := range labels {
+				if label.GetName() != "" && label.GetValue() != "" {
+					tagsMap[label.GetName()] = label.GetValue()
+				}
+			}
+			event.labels = tagsMap
+			event.labelHash = tagsMap.String()
+
+		}
+
+		counter := metric.GetCounter()
+		if counter != nil {
+			value["value"] = int64(counter.GetValue())
+		}
+
+		guage := metric.GetGauge()
+		if guage != nil {
+			value["value"] = guage.GetValue()
+		}
+
+		summary := metric.GetSummary()
+		if summary != nil {
+			value["sum"] = summary.GetSampleSum()
+			value["count"] = summary.GetSampleCount()
+
+			quantiles := summary.GetQuantile()
+
+			percentileMap := common.MapStr{}
+			for _, quantile := range quantiles {
+				key := strconv.FormatFloat((100 * quantile.GetQuantile()), 'f', -1, 64)
+
+				if math.IsNaN(quantile.GetValue()) == false {
+					percentileMap["p"+key] = quantile.GetValue()
```

Comment:
As it is already under `percentile` I don't think we ned to add a `p` in front of the key.